src/Components/Components/src/Forms/InputComponents/InputBase.cs

```diff
@@ -0,0 +1,206 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq.Expressions;
+using System.Threading.Tasks;
+
+namespace Microsoft.AspNetCore.Components.Forms
+{
+    /// <summary>
+    /// A base class for form input components. This base class automatically
+    /// integrates with an <see cref="Forms.EditContext"/>, which must be supplied
+    /// as a cascading parameter.
+    /// </summary>
+    public abstract class InputBase<T> : ComponentBase
+    {
+        private bool _previousParsingAttemptFailed;
+        private ValidationMessageStore _parsingValidationMessages;
+        private Type _nullableUnderlyingType;
+
+        [CascadingParameter] EditContext CascadedEditContext { get; set; }
+
+        /// <summary>
+        /// Gets a value for the component's 'id' attribute.
+        /// </summary>
+        [Parameter] protected string Id { get; private set; }
+
+        /// <summary>
+        /// Gets a value for the component's 'class' attribute.
+        /// </summary>
+        [Parameter] protected string Class { get; private set; }
+
+        /// <summary>
+        /// Gets or sets the value of the input. This should be used with two-way binding.
+        /// </summary>
+        /// <example>
+        /// bind-Value="@model.PropertyName"
+        /// </example>
+        [Parameter] T Value { get; set; }
+
+        /// <summary>
+        /// Gets or sets a callback that updates the bound value.
+        /// </summary>
+        [Parameter] Action<T> ValueChanged { get; set; }
+
+        /// <summary>
+        /// Gets or sets an expression that identifies the bound value.
+        /// </summary>
+        [Parameter] Expression<Func<T>> ValueExpression { get; set; }
```

Comment:
Why do you need both ValueExpression and Value ? it seems redundant. Can you also add some commonly used attributes like "placeholder" or "name"

Reply:
It's much cheaper to evaluate `Value` than `ValueExpression`, so components should only accept a `ValueExpression` if they truly need it.

Reply:
> Can you also add some commonly used attributes like "placeholder" or "name"

This is covered by the desire to pass through arbitrary attributes, which in turn is waiting for support for a "splatting" syntax within the Razor compiler. I've added `class` and `id` because they are so essential; people who want others will have to subclass the built-in `Input*` components in the short term or wait for build-in attribute passthrough in a later preview.